android/clientlib-ktx/src/main/java/com/solana/mobilewalletadapter/clientlib/ActivityResultSender.kt

```diff
@@ -1,7 +1,35 @@
 package com.solana.mobilewalletadapter.clientlib
 
 import android.content.Intent
+import androidx.activity.ComponentActivity
+import androidx.activity.result.ActivityResultLauncher
+import androidx.activity.result.contract.ActivityResultContracts
+import androidx.annotation.GuardedBy
 
-interface ActivityResultSender {
-    fun launch(intent: Intent)
+class ActivityResultSender(
+    rootActivity: ComponentActivity
+) {
+    @GuardedBy("this")
+    private var callback: (() -> Unit)? = null
+
+    private val activityResultLauncher: ActivityResultLauncher<Intent> =
+        rootActivity.registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
+            onActivityComplete()
+        }
+
+    fun startActivityForResult(intent: Intent, onActivityCompleteCallback: () -> Unit) {
+        synchronized(this) {
+            check(callback == null) { "Received an activity start request while another is pending" }
+            callback = onActivityCompleteCallback
+        }
+
+        activityResultLauncher.launch(intent)
```

Comment:
This needs to be wrapped around with activity handler first to verify if the wallet intent could be handled. 

```
        val packageManager = this@MainActivity.packageManager
        if (intent.resolveActivity(packageManager) != null) {
            startForResult.launch(intent)
        } else {
            lifecycleScope.launch(Dispatchers.Main) {
                Toast.makeText(
                    this@MainActivity,
                    "You need to install a Solana Wallet first (Solflare or Phantom)",
                    Toast.LENGTH_LONG
                ).show()
            }
        }
```

If not then we need to return some sort of object as return for failure (maybe a boolean would suffice)

Reply:
At the library level we shouldn't dictate method for notifying the user. BUT - I do think there is a good opportunity either in this library or a companion lib to provide a nice default UI for checking for wallet and presenting some nice default UI. There are two aspects to it:

- This class doesn't crash when there is no destination; returns some kind of status to consumer.
- We create a second, higher level library that has the aforementioned "nice" UI

Reply:
In general though, this is a good catch; I think this would crash as is

Reply:
Yes, it does. Better handling incoming.